fraud/bad_encoding.go

```diff
@@ -0,0 +1,155 @@
+	"github.com/celestiaorg/celestia-node/ipld"
+	ipld_pb "github.com/celestiaorg/celestia-node/ipld/pb"
+	"github.com/celestiaorg/celestia-node/service/header"
+)
+
+type BadEncodingProof struct {
+	BlockHeight uint64
+	// ShareWithProof contains all shares from row or col.
+	// Shares that did not pass verification in rmst2d will be nil.
+	// For non-nil shares MerkleProofs are computed.
+	Shares []*ipld.NamespacedShareWithProof
+	// Index represents the row/col index where ErrByzantineRow/ErrByzantineColl occurred
+	Index uint8
+	// isRow shows that verification failed on row
+	isRow bool
+}
+
+// CreateBadEncodingProof creates a new Bad Encoding Fraud Proof that should be propagated through network
+// Fraud Proof will contain shares that did not pass the verification and appropriate to them Merkle Proofs
+func CreateBadEncodingProof(
+	height uint64,
+	shares *ipld.ErrByzantine,
+) Proof {
+
+	return &BadEncodingProof{
+		BlockHeight: height,
+		Shares:      shares.Shares,
+		isRow:       shares.IsRow,
+		Index:       shares.Index,
+	}
+}
+
+// Type returns type of fraud proof
+func (p *BadEncodingProof) Type() ProofType {
+	return BadEncoding
+}
+
+// Height returns block height
+func (p *BadEncodingProof) Height() uint64 {
+	return p.BlockHeight
+}
+
+// MarshalBinary converts BadEncodingProof to binary
+func (p *BadEncodingProof) MarshalBinary() ([]byte, error) {
+	shares := make([]*ipld_pb.Share, 0, len(p.Shares))
+	for _, share := range p.Shares {
+		shares = append(shares, share.ShareWithProofToProto())
+	}
+
+	badEncodingFraudProof := pb.BadEncoding{
+		Height: p.BlockHeight,
+		Shares: shares,
+		Index:  uint32(p.Index),
+		IsRow:  p.isRow,
+	}
+	return badEncodingFraudProof.Marshal()
+}
+
+// UnmarshalBinary converts binary to BadEncodingProof
+func (p *BadEncodingProof) UnmarshalBinary(data []byte) error {
+	in := pb.BadEncoding{}
+	if err := in.Unmarshal(data); err != nil {
+		return err
+	}
+	befp := UnmarshalBEFP(&in)
+	*p = *befp
+
+	return nil
+}
+
+// UnmarshalBEFP converts given data to BadEncodingProof
+func UnmarshalBEFP(befp *pb.BadEncoding) *BadEncodingProof {
+	return &BadEncodingProof{
+		BlockHeight: befp.Height,
+		Shares:      ipld.ProtoToShare(befp.Shares),
+	}
+}
+
+// Validate ensures that Fraud Proof that was created by full node is correct.
+// Validate checks that provided Merkle Proofs are corresponded to particular shares,
+// rebuilds bad row or col from received shares, computes Merkle Root
+// and compares it with block's Merkle Root
+func (p *BadEncodingProof) Validate(header *header.ExtendedHeader) error {
+	merkleRowRoots := header.DAH.RowsRoots
+	merkleColRoots := header.DAH.ColumnRoots
+	if int(p.Index) >= len(merkleRowRoots) || int(p.Index) >= len(merkleColRoots) {
+		return errors.New("invalid fraud proof: incorrect Index of bad row/col")
+	}
+	if len(merkleRowRoots) != len(merkleColRoots) {
+		return errors.New("invalid fraud proof: invalid extended header")
```

Comment:
nit: in this case it isn't really the fraud proof that is invalid but the user passed in an invalid header only.

Reply:
When I saw this for the first time, I thought that we could even panic in this case as this means sth is seriously wrong here and there is likely no way to recover from this (we should never test a befp against a malformed extended header). For this PR I think we can simply return a better error message and a comment though:
```suggestion
		// NOTE: This should never happen as callers of this method should not feed it with a 
		// malformed extended header. 
		return errors.New("invalid extended header: length of row and column roots do not match")
```

Alternative to the above: add a comment to the method signature that the passed-in extended header had to already pass basic validation and panic here.